[WIP] Add squash and archive history functionality

AnthonyMl opened the pull request:
co-dependent on https://github.com/bdionne/revision-history/pull/4 and https://github.com/bdionne/protege-server/pull/7
TODO:
- remove project names from history and snapshot files

AnthonyMl committed:
Add squash and archive history functionality

- Use ProjectId instead of project name for identifying snapshots
- Check for out of date history snapshot
- Include history snapshot checksum in POST calls in a custom header(only really need it in HTTPChangeService calls).
- Add ProjectId to all HTTPChangeService calls.
- Rewrite POST method to fetch new snapshot and retry on out of date snapshot error code.
- Add progress indicators to squash history and snapshot fetch
- Add setProjectId necessary for tests
- Disable Pause menu on Ontology open if already paused

rgrinberg reviewed (approved):
LGTM. Seems simple enough

bdionne commented:
this looks good. As commented on https://github.com/bdionne/revision-history/pull/4, I'd like to wait and merge all three repos at the same time, as part of a new build

rgrinberg commented on src/main/java/org/protege/editor/owl/client/LocalHttpClient.java:
```diff
@@ -61,10 +70,16 @@
 
+	private final ListeningExecutorService service = MoreExecutors.listeningDecorator(Executors.newSingleThreadExecutor());
+
+	private ProgressDialog dlg = new ProgressDialog();
```
Can we scope these instance variables to the methods where they're being used?

The ProgressDialog initialization fails for some unit tests for example.

bdionne commented on src/main/java/org/protege/editor/owl/client/LocalHttpClient.java:
```diff
@@ -873,15 +874,31 @@ public boolean codeIsLessThan(String lower, String upper) {
+    public ServerStatus getServerStatus() throws ClientRequestException {
+		Response response = null;
+		try {
+			response = get(SERVER_STATUS);
```
This constant is not defined

AnthonyMl replied:
Its in
src/main/java/org/protege/editor/owl/server/http/ServerEndpoints.java
in the protege-server pr

bdionne replied:
oh ok, I probably haven't pulled latest from there

rgrinberg committed:
Add Nonnull in a few places

bdionne committed:
Merge branch 'master' into squash_history

rgrinberg committed:
remove duplicate method

rgrinberg committed:
add missing close brace

rgrinberg committed:
Improved error handling when snapshot missing

AnthonyMl committed:
Manually pass ProjectId to POST calls that require it

rgrinberg committed:
add snapshot check before squashing

rgrinberg commented:
@AnthonyMl Add some `@Nonnull` annotations on the projectId's methods that you've just added/modified.

rgrinberg committed:
projectId input checks

rgrinberg committed:
Return optional when fetching snapshot

rgrinberg committed:
getSnapshot should pass along the projectId to its helper

bdionne commented on src/main/java/org/protege/editor/owl/client/LocalHttpClient.java:
```diff
@@ -467,11 +467,12 @@ public SnapShot getSnapShot(@Nonnull ProjectId projectId) throws LoginTimeoutExc
 	}
 
-	private SnapShot retrieveDocumentSnapshotFromServerResponse(Response response)
+	private SnapShot retrieveDocumentSnapshotFromServerResponse(Response response, @Nonnull ProjectId projectId)
```
this seems odd, it's declared nonnull then there's a check for a null, and either way the variable is never used?

rgrinberg replied:
The boilerplate is there for 2 reasons:

1) Nonnull allows intellij to try and look for places in the code where a null parameter might be passed in.
2) The null check will give a decent error message if it actually occurs at runtime. (These checks can be removed once we're confident that we really aren't passing nulls)

And we are actually using the projectId. It's just that we were erroneously getting it from the instance variable where it was null.

bdionne replied:
? - I think it was an instance variable by design, passing them in is a change

rgrinberg committed:
Get rid of a couple more uses of this.projectId

These methods aren't actually being used currently, but I'm still adding
these parameters so that the instance variable doesn't get reintroduced.

rgrinberg committed:
Use the same projectId in squashHistory

It was using it the id of the project instance which is unneeded

rgrinberg committed:
remove the now unused project instance variable

rgrinberg commented:
The last commit gets rid of that `projectId` local variable entirely. One place where I'd like to see review are those `isWorkflowManager` and `isWorflowModeler` calls. Right now when those client calls fail I'm simply converting them to exceptions. Where the old behavior was to return false. Seems like this is slightly safer. But would like some confirmation there.

bdionne commented:
-0 on this last commit.

rgrinberg committed:
Attempt to get rid of the projectId instance variable

more conservative version

rgrinberg commented:
I addressed the objections to the last commit. The semantics for modeler/manager checking should now be identical. It still seems worrisome to me that we can swallow any client exception in those calls but at least it's safe for now.

Note that this change also requires tweaking nci-edit-tab as well. So I will make that change as well.

AnthonyMl commented on src/main/java/org/protege/editor/owl/client/LocalHttpClient.java:
```diff
@@ -828,12 +828,11 @@ public void putEVSHistory(String code, String name, String operation, String ref
-
-			get(GEN_CON_HIST + "?projectid=" + projectId.get());
-
+			get(GEN_CON_HIST + "?projectid=" + projectId);
```
We need the ```.get()``` here.
toString is different than just the internal id.

rgrinberg committed:
use .get rather than toString

AnthonyMl committed:
Handle update/auto-update server pause failures more gracefully

rgrinberg committed:
Make dialog visible before waiting for future

rgrinberg committed:
Move UI modification to 1 thread

rgrinberg committed:
remove the unused projectId instance variable

AnthonyMl committed:
Move setVisible(false) back into the task

This is necessary as the dialog is modal and setVisible(true) blocks

rgrinberg commented:
closed in favor of #20
